03week/towersOfHanoi.js

```diff
@@ -19,24 +19,104 @@ function printStacks() {
   console.log("c: " + stacks.c);
 }
 
-function movePiece() {
+function reset() {
+  stacks.a = [4, 3, 2, 1];
+  stacks.b = [];
+  stacks.c = [];
+}
+
+function movePiece(source, dest) {
   // Your code here
+  //Check for the legal move
+  //move the last piece of source stack to destination stacks
+  //check for the win
+  if (isLegal(source, dest)) {
+    const piece = stacks[source].pop();
+    stacks[dest].push(piece);
+    checkForWin(dest);
+  } else {
+    console.log('Illegal move, Try again!!')
+  }
 
 }
 
-function isLegal() {
+function isLegal(source, dest) {
   // Your code here
+  //Can move the piece to a destination stack that is empty
+  //check the moving piece against the last piece on the destination stack
+  //should not let to move pieces from a empty stack
+  //source and destination cannot be the same
+  if (stacks[source] === stacks[dest]) {
+    return false;
+  }
+  if (stacks[source].length === 0) {
+    return false;
+  }
+  if (stacks[dest].length === 0) {
+    return true;
+  }
+  let destPiece = stacks[dest].pop();
+  let sourcepiece = stacks[source].pop();
+  if (destPiece > sourcepiece) {
+    stacks[dest].push(destPiece);
```

Comment:
why pop in the first place if the move isn't valid? There is another way that you can get the last item in the array without popping it off.